### PR TITLE
Block import resolution only on 'pub' imports

### LINKED ISSUE #4865: Re-exporting items using a glob breaks if the module imports another module
```rust
priv mod say {
	pub fn hello() { io::println(~"hello"); }
}

pub mod hello {
	use say;

	pub fn hello() {
		say::hello();
	}
}

pub use hello::*; // doesn't work - `pub use hello::hello;` does

fn main() {
	hello();
}
```

Results in this error when compiling:
```
error: failed to resolve imports
a.rs:13:8: 13:17 error: unresolved import
a.rs:13 pub use hello::*; // doesn't work - `pub use hello::hello;` does
                ^~~~~~~~~
a.rs:6:5: 6:9 error: unresolved import
a.rs:6 	use say;
            ^~~~
```

As mentioned in the code, explicitly re-exporting items works but is not a good alternative for when you have hundreds of items (including variants) in a module. Changing the say module to public doesn't fix it.

### PR DESCRIPTION
**Pull request by elinorbgr**
As noted in my previous PR #27439 , the import resolution algorithm has two cases where it bails out:
- The algorithm will delay an import if the module containing the target of the import still has unresolved glob imports
- The algorithm will delay a glob import of the target module still has unresolved imports

This PR alters the behaviour to only bail out when the above described unresolved imports are `pub`, as non-pub imports don't affect the result anyway.

It is still possible to fail the algorithm with examples like

``` rust
pub mod a {
    pub use b::*;
}

pub mod b {
    pub use a::*;
}
```

but such configurations cannot be resolved in any meaningful way, as these are cyclic imports.

Closes #4865

### CODE REVIEW

**Comment by nrc** [comment]
Does this change work in the case of non-public, but visible imports? (i.e., in a nested module). I think it does because glob imports only import public items, not visible ones.

Does this have a backwards compatibility hazard there? I.e., will this prevent us making globs import visible items in the future if we want to?

File: src/test/run-pass/issue-4865-3.rs [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+pub mod a {
+  use b::*;
```
Comment (nrc) [inline src/test/run-pass/issue-4865-3.rs]: nit: indent

File: src/test/run-pass/issue-4865-2.rs [hunk 2]
```diff
@@ -0,0 +1,27 @@
+// Copyright 2015 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
```
Comment (nrc) [inline src/test/run-pass/issue-4865-2.rs]: tests need comments please

File: src/librustc_resolve/build_reduced_graph.rs [hunk 3]
```diff
@@ -926,6 +926,11 @@ impl<'a, 'b:'a, 'tcx:'b> GraphBuilder<'a, 'b, 'tcx> {
                                                                is_public,
                                                                shadowable));
         self.unresolved_imports += 1;
+
+        if is_public {
+            module_.pub_count.set(module_.pub_count.get() + 1);
```
Comment (nrc) [inline src/librustc_resolve/build_reduced_graph.rs]: can you factor this out into an inc_pub_count method?

File: src/librustc_resolve/build_reduced_graph.rs [hunk 4]
```diff
@@ -959,6 +964,9 @@ impl<'a, 'b:'a, 'tcx:'b> GraphBuilder<'a, 'b, 'tcx> {
                 // module's exports ahead of time.
 
                 module_.glob_count.set(module_.glob_count.get() + 1);
+                if is_public {
+                    module_.pub_glob_count.set(module_.pub_glob_count.get() + 1);
```
Comment (nrc) [inline src/librustc_resolve/build_reduced_graph.rs]: likewise

File: src/librustc_resolve/resolve_imports.rs [hunk 5]
```diff
@@ -409,11 +409,19 @@ impl<'a, 'b:'a, 'tcx:'b> ImportResolver<'a, 'b, 'tcx> {
                 GlobImport => {
                     assert!(module_.glob_count.get() >= 1);
                     module_.glob_count.set(module_.glob_count.get() - 1);
+                    if import_directive.is_public {
+                        assert!(module_.pub_glob_count.get() >= 1);
+                        module_.pub_glob_count.set(module_.pub_glob_count.get() - 1);
```
Comment (nrc) [inline src/librustc_resolve/resolve_imports.rs]: and for decrements

File: src/librustc_resolve/resolve_imports.rs [hunk 6]
```diff
@@ -767,16 +775,22 @@ impl<'a, 'b:'a, 'tcx:'b> ImportResolver<'a, 'b, 'tcx> {
...
-        assert_eq!(target_module.glob_count.get(), 0);
-
         // Add all resolved imports from the containing module.
         let import_resolutions = target_module.import_resolutions.borrow();
+
+        if module_.import_resolutions.borrow_state() != ::std::cell::BorrowState::Unused {
+            // In this case, target_module == module_
+            // This means we are trying to glob import a module into itself,
+            // and it is a no-go
+            return ResolveResult::Indeterminate;
```
Comment (nrc) [inline src/librustc_resolve/resolve_imports.rs]: Can we add a specific warning for this case?

**Comment by vberger** [comment]
About visible items, yeah, there might be a compatibility hazard.

This change rely on the fact that item imported by non-`pub` `use` statements are not visible out of the module. This allows to not wait for these import to be resolved to continue, and avoid some deadlock-like situations in the resolution.

However, I will possibly make it harder to introduce help messages in a kind of "maybe you forgot to add `pub` before this import ?". But one could argue that the generic message "There is no `foo` in `bar`." is already quite a good hind about what is wrong.

I'll address your nits tomorrow, it's almost midnight for me. :wink:

**Comment by vberger** [comment]
@nrc For the case of importing a module into itself, is it enough to just add an help message, like this:
```
error: unresolved import `self::*`. cannot glob-import a module into itself [E0432]
```
or should I make a specific error (and add en error code ?)

**Comment by nrc** [comment]
@vberger I don't think you can have different messages for the same error code. Probably the easiest thing to do is to add the extra detail as a note.

**Comment by vberger** [comment]
@nrc well, the error message E0432 already have several messages on it, but they all are a variation on "Unresolved import":
```
error: unresolved import `foo::bar`. Maybe a missing `extern crate foo`? [E0432]
error: unresolved import `foo::bar`. There is no `bar` in `foo`. [E0432]
...
```
The second half of the error message is actually fully-customizable, and I already have the code working to generate the message
```
error: unresolved import `self::*`. Cannot glob-import a module into itself. [E0432]
```
So, would that be fine ?

**Comment by vberger** [comment]
@nrc see my last commit for the implementation

**Comment by nrc** [comment]
The diagnostics stuff just checks the string literal, so if the second part is just `{}` in there then it can change however you like. Otherwise, it needs a different code.

**Comment by nrc** [comment]
lgtm, thanks for the changes!

@bors: r+

**Comment by TimNN** [comment]
Looks like #27619

**Pull request merged**